typescript/radix-sdk/src/utils/query.ts

```diff
@@ -0,0 +1,249 @@
+import {
+  GatewayApiClient,
+  LedgerStateSelector,
+  ScryptoSborValue,
+  StateEntityDetailsResponseItemDetails,
+} from '@radixdlt/babylon-gateway-api-sdk';
+
+import { assert, isNullish, sleep } from '@hyperlane-xyz/utils';
+
+import { EntityDetails, RadixElement } from './types.js';
+
+type RadixComponentDetails = Extract<
+  StateEntityDetailsResponseItemDetails,
+  { type: 'Component' }
+>;
+
+/**
+ * Fetches and validates Radix component details from the gateway.
+ *
+ * @param gateway - The Radix Gateway API client
+ * @param entityAddress - The on-chain address of the component
+ * @param componentName - Human-readable name for error messages (e.g., "mailbox", "ISM", "hook")
+ *
+ * @throws {Error} If the entity is not a Radix component
+ */
+export async function getRadixComponentDetails(
+  gateway: Readonly<GatewayApiClient>,
+  entityAddress: string,
+  componentName: string,
+): Promise<RadixComponentDetails> {
+  const details =
+    await gateway.state.getEntityDetailsVaultAggregated(entityAddress);
+  const componentDetails = details.details;
+
+  assert(
+    componentDetails?.type === 'Component',
+    `Expected on chain details to be defined for radix ${componentName} at address ${entityAddress}`,
+  );
+
+  return componentDetails;
+}
+
+/**
+ * Fetches all keys from a Radix key-value store.
+ *
+ * @param gateway - The Radix Gateway API client
+ * @param keyValueStoreAddress - The on-chain address of the key-value store
+ * @returns Array of all keys in the key-value store
+ */
+export async function getKeysFromKeyValueStore(
+  gateway: Readonly<GatewayApiClient>,
+  keyValueStoreAddress: string,
+): Promise<ScryptoSborValue[]> {
+  let cursor: string | null = null;
+  let at_ledger_state: LedgerStateSelector | null = null;
+  const keys = [];
+  const requestLimit = 50;
+
+  for (let i = 0; i < requestLimit; i++) {
+    const { items, next_cursor, ledger_state } =
+      await gateway.state.innerClient.keyValueStoreKeys({
+        stateKeyValueStoreKeysRequest: {
+          key_value_store_address: keyValueStoreAddress,
+          at_ledger_state,
+          cursor,
+        },
+      });
+
+    keys.push(...items.map((i) => i.key));
+
+    if (!next_cursor) {
+      return keys;
+    }
+
+    cursor = next_cursor;
+    at_ledger_state = { state_version: ledger_state.state_version };
```

Comment:
It may be that the state should remain the same throughout iteration (set it only once, on the first iteration), otherwise we might get inconsistent data.